coderd/workspaceapps/stats.go

```diff
@@ -0,0 +1,378 @@
+package workspaceapps
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/dbauthz"
+)
+
+const (
+	DefaultStatsCollectorReportInterval = 30 * time.Second
+	DefaultStatsCollectorRollupWindow   = 1 * time.Minute
+	DefaultStatsDBReporterBatchSize     = 1024
+)
+
+// StatsReport is a report of a workspace app session.
+type StatsReport struct {
+	UserID           uuid.UUID    `json:"user_id"`
+	WorkspaceID      uuid.UUID    `json:"workspace_id"`
+	AgentID          uuid.UUID    `json:"agent_id"`
+	AccessMethod     AccessMethod `json:"access_method"`
+	SlugOrPort       string       `json:"slug_or_port"`
+	SessionID        uuid.UUID    `json:"session_id"`
+	SessionStartedAt time.Time    `json:"session_started_at"`
+	SessionEndedAt   time.Time    `json:"session_ended_at"` // Updated periodically while app is in use active and when the last connection is closed.
+	Requests         int          `json:"requests"`
+
+	rolledUp bool // Indicates if this report has been rolled up.
+}
+
+func newStatsReportFromSignedToken(token SignedToken) StatsReport {
```

Comment:
If you wanted to you could add `session_id` to the SignedToken to make this easier to corroborate sessions.

Reply:
Wasn't sure if I put it on either SignedToken or SignedToken.Request, will it end up somewhere? Logged or in the DB? I'll defer this to another PR if it makes sense to add.

Reply:
@deansheather does this match what you had in mind? [`357eec2` (#9001)](https://github.com/coder/coder/pull/9001/commits/357eec2b293fba1654ea8589d50297c903c71c8c)

Reply:
Noticed the signed tokens are also created via `Issue`, so the SessionID did not propagate.

I'm afraid my understanding of the signed token lifecycle is too limited to confidently make this change. I added this [`a23297b` (#9001)](https://github.com/coder/coder/pull/9001/commits/a23297b104804c1a3badbe661326bd8df02bf90c), but as a result the session ID will be signed too. Is that something we want or should I add it after the signature is made?

Reply:
I was thinking you should put it in `db.go` where the token is originally created and signed.

Reply:
Yeah I think it's fine if it's included in the signed payload. Means it persists between requests for up to a minute right?

Reply:
If you don't need/want that feel free to dismiss this though

Reply:
IIRC there is a certain token re-use, will there be one token created per request or will the same one (and thus, sessionID) be re-used across multiple requests?

If it's the latter, it will conflict with the rollup logic I believe.

Reply:
Chatted with Dean on Discord and we decided to not try to corroborate requests/tokens for now. Since SignedTokens live for one minute and are re-used, re-using that ID would conflict with rollups of short and long-lived connections.